api/data_test.go

```diff
@@ -149,3 +149,88 @@ func TestCreateNewCacheTime(t *testing.T) {
+
+		So(responseRecorder.Code, ShouldEqual, 400)
+		So(responseRecorder.Body.String(), ShouldContainSubstring, "empty request body")
+	})
+
+	Convey("When an etag and/or path is not provided, 400 bad request is returned", t, func() {
+		ctx := context.Background()
+		dataStoreMock := &mock.DataStoreMock{}
+		staticTimeString := staticTime.Format(time.RFC3339)
+		body := `{"collection_id": 123, "release_time":"` + staticTimeString + `"}`
+
+		request := httptest.NewRequest(http.MethodPut, baseURL+testCacheID, bytes.NewBufferString(body))
+		responseRecorder := httptest.NewRecorder()
+		dataStoreAPI := setupAPIWithStore(ctx, dataStoreMock)
+		dataStoreAPI.Router.ServeHTTP(responseRecorder, request)
+
+		So(responseRecorder.Code, ShouldEqual, 400)
+		So(responseRecorder.Body.String(), ShouldContainSubstring, "[etag field missing path field missing]")
+	})
+
+	Convey("When an extra field is provided, a 400 bad request is returned", t, func() {
+		ctx := context.Background()
+		dataStoreMock := &mock.DataStoreMock{}
+		body := `{"path": "testpath", "etag": "testetag", "extra_field": "hello" }`
+
+		request := httptest.NewRequest(http.MethodPut, baseURL+testCacheID, bytes.NewBufferString(body))
+		responseRecorder := httptest.NewRecorder()
+		dataStoreAPI := setupAPIWithStore(ctx, dataStoreMock)
+		dataStoreAPI.Router.ServeHTTP(responseRecorder, request)
+
+		So(responseRecorder.Code, ShouldEqual, 400)
+		So(responseRecorder.Body.String(), ShouldContainSubstring, `json: unknown field "extra_field"`)
+	})
+
+	Convey("When the field type provided is not the expected, a 400 bad request is returned", t, func() {
+		ctx := context.Background()
+		dataStoreMock := &mock.DataStoreMock{}
+		body := `{"path": 1234, "etag": "testetag"}`
+		request := httptest.NewRequest(http.MethodPut, baseURL+testCacheID, bytes.NewBufferString(body))
+		responseRecorder := httptest.NewRecorder()
+		dataStoreAPI := setupAPIWithStore(ctx, dataStoreMock)
+		dataStoreAPI.Router.ServeHTTP(responseRecorder, request)
+
+		So(responseRecorder.Code, ShouldEqual, 400)
+		So(responseRecorder.Body.String(), ShouldContainSubstring, "json: cannot unmarshal number into Go struct field CacheTime.path of type string")
+	})
+
+	Convey("When the id provided is not 32 characters in length, a 400 bad request is returned", t, func() {
+		ctx := context.Background()
+		dataStoreMock := &mock.DataStoreMock{}
+		body := validBody
+		request := httptest.NewRequest(http.MethodPut, baseURL+"abc", bytes.NewBufferString(body))
+		responseRecorder := httptest.NewRecorder()
+		dataStoreAPI := setupAPIWithStore(ctx, dataStoreMock)
+		dataStoreAPI.Router.ServeHTTP(responseRecorder, request)
+
+		So(responseRecorder.Code, ShouldEqual, 400)
+		So(responseRecorder.Body.String(), ShouldContainSubstring, "[id should be 32 characters in length]")
+	})
+
+	Convey("When the id provided is 32 characters but not hexadecimal lowercase, a 400 bad request is returned", t, func() {
```

Comment:
This should be 2 tests: one testing for non-hexadecimal, another one for non-lowercase.

Testing just one thing per test helps to isolate what you're testing, and it makes it easier to identify what failed when something breaks.

Reply:
Good point, we had created a function to check both of these things at the same time with a regular expression, I will update this to two separate checks.

Reply:
Just to clarify, you can still just use a single regular expression and check for the ID being hexadecimal and lowercase at the same time if you think that's what makes more sense, but there should definitely be two separate tests. If you use the Given When Then style, this could look something like this:
```
Given an upsert request
  When an ID with uppercase characters is provided
    Then a 400 error is thrown
  When an ID with non-hexadecimal characters is provided
    Then a 400 error is thrown
```